### PR TITLE
Add Cache-Control headers for catalog leaves and finished pages

### PR DESCRIPTION
**Pull request by joelverhagen**
Implements spec: https://github.com/NuGet/Engineering/pull/4876
Progress on https://github.com/NuGet/Engineering/issues/4875:

- Adds `-itemCacheControl` and `-finishedPageCacheControl` options to Db2Catalog, defaults to `no-store` (current behavior)
- Set all catalog leaf documents to have the specified cache control value
- When a catalog page is added set the specified cache control header on the previous page
- Fix flaky Db2Catalog tests (I think). They were using a non-seeded `Random` and current time.
- Deleted unused `ICatalogGraphPersistence`.
- Added `-storageInitializeContainer` and `-sendX5c` options, which help in local debugging
  - `-storageInitializeContainer` allows you to prevent the storage abstraction from creating already created containers (allows a read-only SAS token)
  - `-sendX5c` allows using SN+I based certificate authentication for KeyVault.

I will be running this on DEV for a while to make sure it works okay.

### CODE REVIEW

File: src/Catalog/Persistence/AzureStorage.cs [hunk 1]
```diff
@@ -182,6 +182,34 @@ public override async Task<IEnumerable<StorageListItem>> ListAsync(CancellationT
...
+
+            if (blob.Properties.CacheControl != cacheControl)
+            {
+                blob.Properties.CacheControl = cacheControl;
+
+                await blob.SetPropertiesAsync(
+                    accessCondition: AccessCondition.GenerateIfMatchCondition(blob.Properties.ETag),
+                    options: null,
+                    operationContext: null,
+                    cancellationToken);
```
Comment (theot12) [inline src/Catalog/Persistence/AzureStorage.cs:204]: Why is this line different from line 194? 

```suggestion
                    cancellationToken: cancellationToken);
```

File: tests/NgTests/Db2CatalogTests.cs [hunk 2]
```diff
@@ -1058,6 +1247,8 @@ public async Task RunInternal_CallsCatalogStorageLoadStringExactlyOnce()
                 top: top,
                 verbose: true,
                 Constants.MaxPageSize,
+                Constants.NoStoreCacheControl,
+                Constants.NoStoreCacheControl,
```
Comment (theot12) [inline tests/NgTests/Db2CatalogTests.cs:1251]: Consider changing this to add clarity. 

```suggestion
                itemCacheControl: Constants.NoStoreCacheControl,
                finishedPageCacheControl: Constants.NoStoreCacheControl,
```

File: tests/NgTests/Db2CatalogTests.cs [hunk 3]
```diff
@@ -1058,6 +1247,8 @@ public async Task RunInternal_CallsCatalogStorageLoadStringExactlyOnce()
                 top: top,
                 verbose: true,
                 Constants.MaxPageSize,
```
Comment (theot12) [inline tests/NgTests/Db2CatalogTests.cs:1249]:
```suggestion
                maxPageSize: Constants.MaxPageSize,
```

File: tests/NgTests/Db2CatalogTests.cs [hunk 4]
```diff
@@ -1161,6 +1354,8 @@ public async Task RunInternal_CallsCatalogStorageLoadStringExactlyOnce()
                 top: top,
                 verbose: true,
                 maxPageSize,
+                itemCacheControl,
+                finishedPageCacheControl,
```
Comment (theot12) [inline tests/NgTests/Db2CatalogTests.cs:1358]: Consider changing to this for consistency throughout. 

```suggestion
                maxPageSize: Constants.MaxPageSize,
                itemCacheControl: Constants.NoStoreCacheControl,
                finishedPageCacheControl: Constants.NoStoreCacheControl,
```

File: tests/NgTests/Db2CatalogTests.cs [hunk 5]
```diff
@@ -1147,7 +1338,9 @@ public async Task RunInternal_CallsCatalogStorageLoadStringExactlyOnce()
             bool skipCreatedPackagesProcessing,
             int top,
             Mock<IGalleryDatabaseQueryService> galleryDatabaseMock,
-            int maxPageSize = Constants.MaxPageSize)
+            int maxPageSize = Constants.MaxPageSize,
+            string itemCacheControl = Constants.NoStoreCacheControl,
+            string finishedPageCacheControl = Constants.NoStoreCacheControl)
```
Comment (theot12) [inline tests/NgTests/Db2CatalogTests.cs:1343]: consistency. see below.

File: tests/NgTests/Db2CatalogTests.cs [hunk 6]
```diff
@@ -705,6 +888,8 @@ public async Task CreatesNewCatalogFromCreatedAndEditedAndDeletedPackages()
                 top: top,
                 verbose: true,
                 Constants.MaxPageSize,
+                Constants.NoStoreCacheControl,
+                Constants.NoStoreCacheControl,
```
Comment (theot12) [inline tests/NgTests/Db2CatalogTests.cs:892]: consistency, see below.

File: tests/NgTests/Db2CatalogTests.cs [hunk 7]
```diff
@@ -826,6 +1011,8 @@ public async Task AppendsDeleteToExistingCatalog()
...
```
Comment (theot12) [inline tests/NgTests/Db2CatalogTests.cs:1015]: consistency, see below.

File: tests/NgTests/Db2CatalogTests.cs [hunk 8]
```diff
@@ -944,6 +1131,8 @@ public async Task AppendsDeleteAndReinsertToExistingCatalog()
...
```
Comment (theot12) [inline tests/NgTests/Db2CatalogTests.cs:1135]: consistency, see below.

**Review by theot12 — APPROVED** [review]

**Comment by zhhyu** [comment]
Looking ...

File: tests/NgTests/Db2CatalogTests.cs (hunk 4 above)
Comment (agr) [reply tests/NgTests/Db2CatalogTests.cs:1358]: `maxPageSize: Constants.MaxPageSize` looks too verbose to me. The intent is obvious from the constant name, no need to make it even more explicit with parameter name. The other two though are a good clarification.

File: src/Catalog/AppendOnlyCatalogWriter.cs [hunk 9]
```diff
@@ -81,29 +67,37 @@ protected override Uri[] GetAdditionalRootType()
...
             pageEntries[pageUri.AbsoluteUri] = new CatalogItemSummary(Schema.DataTypes.CatalogPage, commitId, commitTimeStamp, items.Count);
 
-            return pageEntries;
+            return new SavePagesResult(pageEntries, finishedPageUri);
         }
 
-        private Uri GetPageUri(IDictionary<string, CatalogItemSummary> currentPageEntries, int newItemCount, out bool isExistingPage)
+        private Uri GetPageUri(IDictionary<string, CatalogItemSummary> currentPageEntries, int newItemCount, out bool isExistingPage, out Uri previousPageUri)
         {
             Tuple<int, Uri, int> latest = ExtractLatest(currentPageEntries);
```
Comment (agr) [inline src/Catalog/AppendOnlyCatalogWriter.cs:84]: Should we use tuple deconstruct [syntax](https://learn.microsoft.com/en-us/dotnet/csharp/fundamentals/functional/deconstruct#tuples) here? Assuming we have compiler support.

File: src/Catalog/CatalogWriterBase.cs [hunk 10]
```diff
@@ -16,17 +16,18 @@ namespace NuGet.Services.Metadata.Catalog
...
-        public CatalogWriterBase(IStorage storage, ICatalogGraphPersistence graphPersistence = null, CatalogContext context = null)
+        public CatalogWriterBase(IStorage storage, ITelemetryService telemetryService, CatalogContext context = null)
         {
             Options.InternUris = false;
 
             Storage = storage;
-            GraphPersistence = graphPersistence;
             Context = context ?? new CatalogContext();
 
+            _telemetryService = telemetryService;
```
Comment (agr) [inline src/Catalog/CatalogWriterBase.cs:30]: nul check?

File: src/Catalog/Persistence/AzureStorage.cs (hunk 1 above)
Comment (agr) [reply src/Catalog/Persistence/AzureStorage.cs:204]: I suggest we remove repetition from line 194 instead (if compiler is happy, of course).

File: src/Catalog/Persistence/AzureStorage.cs [hunk 11]
```diff
@@ -182,6 +182,34 @@ public override async Task<IEnumerable<StorageListItem>> ListAsync(CancellationT
...
+                return true;
+            }
+            else
```
Comment (agr) [inline src/Catalog/Persistence/AzureStorage.cs:207]: This `else` block is unnecessary. Just `return false;` will do the same.

File: tests/NgTests/Db2CatalogTests.cs (hunk 4 above)
Comment (agr) [reply tests/NgTests/Db2CatalogTests.cs:1358]: Wait, the whole suggestion changes the code behavior: original code passes function arguments here, suggested change discards those arguments and passes constants.

**Review by agr — DISMISSED** [review]

**Review by zhhyu — DISMISSED** [review]

File: tests/NgTests/Db2CatalogTests.cs (hunk 6 above)
Comment (joelverhagen) [reply tests/NgTests/Db2CatalogTests.cs]: Fixed.

File: tests/NgTests/Db2CatalogTests.cs (hunk 8 above)
Comment (joelverhagen) [reply tests/NgTests/Db2CatalogTests.cs]: Fixed.

File: src/Catalog/AppendOnlyCatalogWriter.cs (hunk 9 above)
Comment (joelverhagen) [reply src/Catalog/AppendOnlyCatalogWriter.cs]: Great idea! I didn't think it was supported but it is!

File: src/Catalog/CatalogWriterBase.cs (hunk 10 above)
Comment (joelverhagen) [reply src/Catalog/CatalogWriterBase.cs]: Fixed.

File: src/Catalog/Persistence/AzureStorage.cs (hunk 11 above)
Comment (joelverhagen) [reply src/Catalog/Persistence/AzureStorage.cs]: Fixed.

File: tests/NgTests/Db2CatalogTests.cs (hunk 7 above)
Comment (joelverhagen) [reply tests/NgTests/Db2CatalogTests.cs]: Fixed.

File: tests/NgTests/Db2CatalogTests.cs (hunk 5 above)
Comment (joelverhagen) [reply tests/NgTests/Db2CatalogTests.cs:1343]: Fixed.

File: tests/NgTests/Db2CatalogTests.cs (hunk 4 above)
Comment (joelverhagen) [reply tests/NgTests/Db2CatalogTests.cs:1358]: I left the code as-is. The variable names are descriptive enough.

File: tests/NgTests/Db2CatalogTests.cs (hunk 3 above)
Comment (joelverhagen) [reply tests/NgTests/Db2CatalogTests.cs]: Fixed.

File: tests/NgTests/Db2CatalogTests.cs (hunk 2 above)
Comment (joelverhagen) [reply tests/NgTests/Db2CatalogTests.cs]: Fixed.

File: src/Catalog/Persistence/AzureStorage.cs (hunk 1 above)
Comment (joelverhagen) [reply src/Catalog/Persistence/AzureStorage.cs:204]: I removed the named parameter in these cases since it's clear.

**Review by agr — APPROVED** [review]

**Pull request merged**